touchforms/formplayer/static/formplayer/script/fullform-ui.js

```diff
@@ -1,634 +1,459 @@
-    if (this.$error) {
-        this.$error.text(content ? content : '');
     }
-    this.$container[content ? 'addClass' : 'removeClass']('qerr');
-  }
 
-  this.clearError = function() {
-    this.showError(null);
-  }
-
-
-  this.destroy = function() {
-    this.control.destroy();
-  }
-}
+    self.showError = function(content) {
+        self.error(content);
+    };
 
-function make_element(e, parent) {
-  if (e.type == 'question') {
-    var o = new Question(e, parent);
-  } else if (e.type == 'sub-group') {
-    var o = new Group(e, parent);
-  } else if (e.type == 'repeat-juncture') {
-    var o = new Repeat(e, parent);
-  }
-  o.init_render();
-  return o;
-}
+    self.clearError = function() {
+        self.error(null);
+    };
 
-function render_elements(parent, elems) {
-  for (var i = 0; i < elems.length; i++) {
-    var o = make_element(elems[i], parent);
-    parent.children.push(o);
-    parent.child_container().append(o.$container);
-  }
-  empty_check(parent);
 }
 
-function reconcile_elements(parent, new_elems) {
-  var mapping = [];
-  for (var i = 0; i < parent.children.length; i++) {
-    var child = parent.children[i];
-    mapping.push([child, inElementSet(child, new_elems)]);
-  }
-  for (var i = 0; i < new_elems.length; i++) {
-    var new_elem = new_elems[i];
-    if (inElementSet(new_elem, parent.children) == null) {
-      mapping.push([null, new_elem]);
+/**
+ * Reconciles the JSON representation of a Question and renders it into
+ * a knockout representation.
+ * @param {Object} json - The JSON returned from touchforms to represent a Question
+ */
+Question.prototype.fromJS = function(json) {
+    var self = this;
+    var mapping = {
+        caption: {
+            update: function(options) {
+                return options.data ? DOMPurify.sanitize(options.data.replace(/\n/g, '<br/>')) : null;
+            }
+        },
+        caption_markdown: {
+            update: function(options) {
+                return options.data ? markdowner.render(options.data) : null;
+            }
+        },
     }
-  }
 
-  $.each(mapping, function(i, val) {
-      var e_old = val[0];
-      var e_new = val[1];
-
-      if (e_old == null) {
-        var o = make_element(e_new, parent);
-        addChild(parent, o, new_elems);
-      } else if (e_new == null) {
-        deleteChild(parent, e_old);
-      } else {
-        e_old.reconcile(e_new);
-      }
-    });
+    if (json.datatype === Formplayer.Const.MULTI_SELECT) {
+        json.answer = _.map(json.answer, function(d) { return '' + d });
+    }
+    ko.mapping.fromJS(json, mapping, self);
 }
 
-function deleteChild(parent, child) {
-  child.$container.slideUp('slow', function() {
-      child.$container.remove();
-      child.destroy();
-      arrayDelItem(parent.children, child);
-      if (window.mainView) {
-        window.mainView.router.view.dirty = true
-      }
-      empty_check(parent, 'fast');
+function reconcileElements(parent, new_elems) {
```

Comment:
This was a huge change. we were doing this pretty complicated reconciliation of questions and that all got very simple by setting up the proper `ko.mapping.fromJS`